tiledb/array.py

```diff
@@ -243,3 +251,748 @@ def index_domain_subarray(array, dom, idx: tuple):
+    """Base class for TileDB array objects.
+
+    Defines common properties/functionality for the different array types. When
+    an Array instance is initialized, the array is opened with the specified mode.
+
+    :param str uri: URI of array to open
+    :param str mode: (default 'r') Open the array object in read 'r', write 'w', modify exclusive 'm', or delete 'd' mode
+    :param str key: (default None) If not None, encryption key to decrypt the array
+    :param tuple timestamp: (default None) If int, open the array at a given TileDB
+        timestamp. If tuple, open at the given start and end TileDB timestamps.
+    :param str attr: (default None) open one attribute of the array; indexing a
+        dense array will return a Numpy ndarray directly rather than a dictionary.
+    :param Ctx ctx: TileDB context
+    """
+
+    def __init__(
+        self, uri, mode="r", key=None, timestamp=None, attr=None, ctx=None, **kwargs
+    ):
+        if ctx is None:
+            ctx = default_ctx()
+
+        self.array = kwargs.get(
+            "preloaded_array", preload_array(uri, mode, key, timestamp, ctx)
+        )
+
+        # view on a single attribute
+        schema = self.array._schema()
+        if attr is not None and not schema._has_attribute(attr):
+            self.array._close()
+            raise KeyError(f"No attribute matching '{attr}'")
+        else:
+            self.view_attr = attr
+
+        self.ctx = ctx
+        self.key = key
+        self.uri = uri
+        self.domain_index = DomainIndexer(self)
+        self.pyquery = None
+        self.__buffers = None
+
+        self.last_fragment_info = dict()
+        self.meta = Metadata(self)
+
+    def __capsule__(self):
+        return self.array.__capsule__()
+
+    def __repr__(self):
+        if self.isopen:
+            array_type = "Sparse" if self.schema.sparse else "Dense"
+            return f"Array(type={array_type }, uri={self.uri!r}, mode={self.mode}, ndim={self.schema.ndim})"
+        else:
+            return f"Array(uri={self.uri!r}, mode=closed)"
+
+    @classmethod
+    def create(cls, uri, schema, key=None, overwrite=False, ctx=None):
+        """Creates a TileDB Array at the given URI
+
+        :param str uri: URI at which to create the new empty array.
+        :param ArraySchema schema: Schema for the array
+        :param str key: (default None) Encryption key to use for array
+        :param bool overwrite: (default False) Overwrite the array if it already exists
+        :param Ctx ctx: (default None) Optional TileDB Ctx used when creating the array,
+                        by default uses the ArraySchema's associated context
+                        (*not* necessarily ``tiledb.default_ctx``).
+
+        """
+        if ctx is not None and not isinstance(ctx, Ctx):
+            raise TypeError(
+                "tiledb.Array.create() expected tiledb.Ctx " "object to argument ctx"
+            )
+        if ctx is None:
+            ctx = schema.ctx
```

Comment:
```suggestion
        ctx = ctx or default_ctx()
```

We should not be using the internal context of "data objects" like array schemas.